parse quote offline to save gas

smtmfft opened the pull request:
Current quote calldata is pure bytes and parsing bytes is costy. 
Moving most of parse logic offline saves ~1M.
See verifyAttestation vs verifyParsedQuote below:

Running 4 tests for forge-test/AutomataDcapV3AttestationTest.t.sol:AutomataDcapV3AttestationTest
[PASS] testAttestation() (gas: 3900827)
[PASS] testCRL() (gas: 2411171)
[PASS] testCRLWithParsedQuote() (gas: 2456327)
[PASS] testParsedQuoteAttestation() (gas: 3970356)
Test result: ok. 4 passed; 0 failed; 0 skipped; finished in 44.93ms
|AutomataDcapV3Attestation |                 |         |         |         |         |
|----------------------------------------------------------------------------|-----------------|---------|---------|---------|---------|
| Function Name                                                              | min             | avg     | median  | max     | # calls |
| verifyAttestation                                                          | 2049387         | 2808334 | 2808334 | 3567281 | 2       |
| verifyParsedQuote                                                          | 999962          | 1770977 | 1770977 | 2541992 | 2       |

smtmfft committed:
parse quote offline to save gas

Signed-off-by: smtmfft <smtm@taiko.xyz>

pbeza commented:
We should probably deduplicate copy-pasted code from:
https://github.com/automata-network/automata-dcap-v3-attestation/blob/c59775ada150949660c5dda6d1f83616504b413d/contracts/AutomataDcapV3Attestation.sol#L394

and:

https://github.com/automata-network/automata-dcap-v3-attestation/blob/c59775ada150949660c5dda6d1f83616504b413d/contracts/AutomataDcapV3Attestation.sol#L129
(unless we remove the less performant version and keep the other one)

smtmfft commented:
> We should probably deduplicate copy-pasted code from:
> 
> https://github.com/automata-network/automata-dcap-v3-attestation/blob/c59775ada150949660c5dda6d1f83616504b413d/contracts/AutomataDcapV3Attestation.sol#L394
> 
> and:
> 
> https://github.com/automata-network/automata-dcap-v3-attestation/blob/c59775ada150949660c5dda6d1f83616504b413d/contracts/AutomataDcapV3Attestation.sol#L129
> 
> 
> (unless we remove the less performant version and keep the other one)

Yes, if this approach is ok, I think we can remove the previous one. so far the old one serves as a implement reference and perf comparison base.

melynx commented:
hey @smtmfft , sorry for the late response. the approach proposed is actually something that we are looking to implement! thank you so much for the contribution. 

@preston4896 is looking through the code and if everything is ok we'll merge it in :).

preston4896 commented on contracts/lib/QuoteV3Auth/V3Struct.sol:
```diff
@@ -33,6 +33,11 @@ library V3Struct {
         bytes data;
     }
 
+    struct ParsedQEAuthData {
```
This structure is redundant. We can simply store the data size as a uint16 in the original QEAuthData struct. The gas conversion from uint256 to uint16 is negligible.

smtmfft replied:
Thanks! NP, I will change this soon. I did not change previous struct is to avoid change code logic as I want to compare the result with exactly the previous path.

another question is: is that safe to change TcbInfo member from uint256 to uint8? https://github.com/automata-network/automata-dcap-v3-attestation/blob/e731c5b532e3b080daf333bac3af405ab20ea30e/contracts/lib/TCBInfoStruct.sol#L11-L15.
As I get response from intel, also from the tcb.json file, those value are less equal than 255. Thanks!

preston4896 replied:
I wouldn't worry about this for now, as long as it doesn't affect functionality (not breaking) :)

preston4896 commented on contracts/lib/QuoteV3Auth/V3Struct.sol:
```diff
@@ -41,6 +46,13 @@ library V3Struct {
         bytes certData;
     }
 
+    struct ParsedCertificationData {
```
Also redundant, but let's store certType as a uint16, and certDataSize as a uint32.

For the cert data array however, I do not recommend storing them as a fixed length at 3. As seen from the Intel source code at https://github.com/intel/SGX-TDX-DCAP-QuoteVerificationLibrary/blob/16b7291a7a86e486fdfcf1dfb4be885c0cc00b4e/Src/AttestationLibrary/src/QuoteVerification/QuoteStructures.h#L135-L141, they are stored as a vector instead.

This essentially provides us with the flexibility to store a cert chain of various size, if Int

preston4896 commented on contracts/lib/QuoteV3Auth/V3Struct.sol:
```diff
@@ -41,6 +46,13 @@ library V3Struct {
         bytes certData;
     }
 
+    struct ParsedCertificationData {
```
Same redundancy and reasoning as the previous Parsed Struct. But we can retain:
- `uint16 certType`
- `uint32 certDataSzie`
- `bytes[3] certData` (because this struct is strictly used for V3 quotes, and i don't anticipate Intel to introduce a PCK cert chain of different length for this version)

smtmfft replied:
Yes, so far it's V3 ECDSA certification chain (certType==5) specific, but I guess it's ok for now.

preston4896 commented on contracts/lib/QuoteV3Auth/V3Struct.sol:
```diff
@@ -49,4 +61,19 @@ library V3Struct {
         CertificationData certification;
     }
+
+    struct ParsedECDSAQuoteV3AuthData {
```
Redundancy also. And no need to store qe report as raw for the original struct. (consider removing line 161 from `AutomataDcapV3Attestation.sol`

smtmfft replied:
removed

preston4896 commented on contracts/lib/QuoteV3Auth/V3Struct.sol:
```diff
@@ -49,4 +61,19 @@ library V3Struct {
+        ParsedCertificationData certification;
+    }
+
+    struct ParsedV3QuoteStruct {
```
Keeping this, for `verifyParsedQuote()`

smtmfft replied:
Yes, I keep these 2 entries, one for bytes quote and one for parsed bytes quote, and they share the same verifyParsedQuote logic.

preston4896 commented on contracts/AutomataDcapV3Attestation.sol:
```diff
@@ -373,4 +373,166 @@ contract AutomataDcapV3Attestation is IAttestation {
+        return (_attestationTcbIsValid(tcbStatus), retData);
+    }
+
+    function _enclaveParsedReportSigVerification(
```
Line 503-537 is a duplicate. Remove entirely.

smtmfft replied:
removed

preston4896 commented on contracts/AutomataDcapV3Attestation.sol:
```diff
@@ -373,4 +373,166 @@ contract AutomataDcapV3Attestation is IAttestation {
+    /// @dev For all valid quote verification, returns the following data:
+    /// (_attestationTcbIsValid())
+    /// @dev exitCode is defined in the {{ TCBInfoStruct.TCBStatus }} enum
+    function verifyParsedQuote(V3Struct.ParsedV3QuoteStruct calldata v3quote) external returns (bool, bytes memory) {
```
Consider putting step 2 - 8 into it's own code block (internal function), and it can also be used by the internal `_verify()` method

smtmfft replied:
Just make the _verifyParsedQuote a common function instead, as I think these steps are clear enough for now, and also each step is not big in code line.

preston4896 commented on contracts/lib/QuoteV3Auth/V3Parser.sol:
```diff
@@ -154,4 +154,120 @@ library V3Parser {
+        //     "Invalid certData size"
+        // );
+
+        uint32 totalQuoteSize = 48 // header
```
Minimum quote length is defined [here](https://github.com/intel/SGX-TDX-DCAP-QuoteVerificationLibrary/blob/16b7291a7a86e486fdfcf1dfb4be885c0cc00b4e/Src/AttestationLibrary/src/QuoteVerification/QuoteConstants.h#L95-L99)

it does not take the sizes of authData and certData values itself into account, but rather the bytes allocated to store `authDataSize` (2 bytes), `certDataSize` (4 bytes) and `certDataType` (2 bytes).

The auth_data_size_byte_len is also missing here which is 4 bytes.

smtmfft replied:
Oh, yes, missed few sizeof(DataSize)

preston4896 replied:
Consider removing this length check completely, because the numbers here adds up to exactly `MINIMUM_QUOTE_LENGTH` - 4 . 

`MINIMUM_QUOTE_LENGTH` is more appropriately used for checking quotes that are sent in raw bytes form to ensure it is conform to the proper byte length defined in the various Quote structs.

Since the input to `verifyParsedQuote()` is a Solidity struct, most values are type-safe and can safely assume to be in the correct byte length. Therefore, we only have to check for variables that are of dynamically-lengthed `bytes` type, such as `qeReport.reserved3`, `qeReport.reserved4`, ECDSA key and signature.

I am working on these changes now, and will make a staging commit that will eventually be merged into `main`.

preston4896 commented on contracts/lib/QuoteV3Auth/V3Parser.sol:
```diff
@@ -154,4 +154,120 @@ library V3Parser {
+            "Invalid QEAuthData size"
+        );
+
+        // todo!
```
TODO: We should look into the gas cost to perform the base64 encoding on the individual certificates here for the length check.

preston4896 replied:
I think we can make this as a separate functionality to work on, since the scope of this PR is strictly on parsing the quote.

preston4896 reviewed (commented):
I have added a few comments on specific lines of the code where I think we can improve upon. But overall, I think it's great to have the option to not parse the entire quote on-chain.

smtmfft committed:
Code refine according to review comments.

1. clean redundent members.
2. make verifyQuote & verifyParsedQuote share the same verification logic which is based on parsedQuote.
3. fix some issues.

preston4896 committed:
minor updates and code formatting

preston4896 commented:
Hey @smtmfft, I have just pushed a few changes directly to this PR, to address the following:

- Removed `MINIMUM_QUOTE_LENGTH` check from the `validateParsedInput()` method in `V3Parser.sol`
- Optimize the `_verifyParsedQuote()` common function in `AutomataDcapV3Attestation.sol`, such that it will only validate the input if the quote were parsed off-chain. Since, we can trust the quote that are parsed directly on-chain to be in the correct format. 
- Minor syntax error in `hardhat.config.ts`
- Unit tests should stick to running on a fork, to the save the steps from re-deploying `P256Verifier` library.

The [commit](https://github.com/automata-network/automata-dcap-v3-attestation/pull/6/commits/018cfb4bda1d15a3b73cd845794d66cda86b6b11) shall serve as a "final call" before I merge to main. Please feel free to lmk if you have any questions. :)

preston4896 committed:
remove p256verifier from test
